### ISSUE TITLE
Inconsistent justification

### ISSUE THREAD

**Issue by mbossenbroek** [kkinnear/zprint#48]
Given the following example and settings, I'm getting an unexpected result. The first map comes out as expected, but the second one is not justified. Do you know what would cause this?

(apologies for the confusing example - I tried for a simpler repro case, but obfuscating the data was easiest)

Input
```
(zp/zprint-str
  '[{:abcd         :abcd
     :abcdef       {:abcdefg  [:abcdefgh :abcdefghijkl :abcdefgh :850fda697252 variable]
                    :abcdefgh [:s3 :abcdefgh-ijklmn :abcdefghijkl :abcdefgh :850fda697252 variable]}
     :abcdefghijkl []
     :abce         {:abcd   :abcd
                    :abcdef {:abcd "foo"}}}
    {:abcd         :abcd
     :abcdef       {:abcdefg  [:abcdefgh :abcdefghijkl :850fda697252 variable]
                    :abcdefgh [:ab :abcdefgh-ijklmn :abcdefghijkl :850fda697252 variable]}
     :abcdefghijkl [{:abcd    "bar"
                     :abcdefg [:abcdefgh :abcdefghijkl :abcdefgh :850fda697252 variable]}]
     :abce         {:abcd   :abcd
                    :abcdef {:abcd "foo"}}}]
  100 {:vector {:wrap? false}
       :map    {:comma?    false
                :force-nl? true
                :justify?  true}})
```

Result:
```
[{:abcd         :abcd
  :abcdef       {:abcdefg  [:abcdefgh :abcdefghijkl :abcdefgh :850fda697252 variable]
                 :abcdefgh [:s3 :abcdefgh-ijklmn :abcdefghijkl :abcdefgh :850fda697252 variable]}
  :abcdefghijkl []
  :abce         {:abcd   :abcd
                 :abcdef {:abcd "foo"}}}
 {:abcd :abcd
  :abcdef {:abcdefg  [:abcdefgh :abcdefghijkl :850fda697252 variable]
           :abcdefgh [:ab :abcdefgh-ijklmn :abcdefghijkl :850fda697252 variable]}
  :abcdefghijkl
    [{:abcd "bar"
      :abcdefg
        [:abcdefgh :abcdefghijkl :abcdefgh :850fda697252 variable]}]
  :abce {:abcd   :abcd
         :abcdef {:abcd "foo"}}}]
```

**Comment by kkinnear** [comment]
Wow, I thought that this would be another case of -- it won't justify if the justification is longer than the unjustified version.  But it is, instead, a bug (for which I don't yet have a fix).  For instance, the map that didn't justify will justify if it isn't in a vector, but not if it is:
```clojure
(czprint i48z {:width 100 :map {:comma? false :force-nl? true :justify? true} :vector {:wrap? false}})

{:abcd         :abcd
 :abcdef       {:abcdefg  [:abcdefgh :abcdefghijkl :850fda697252 variable]
                :abcdefgh [:ab :abcdefgh-ijklmn :abcdefghijkl :850fda697252 variable]}
 :abcdefghijkl [{:abcd    "bar"
                 :abcdefg [:abcdefgh :abcdefghijkl :abcdefgh :850fda697252 variable]}]
 :abce         {:abcd   :abcd
                :abcdef {:abcd "foo"}}}

(czprint [i48z] {:width 100 :map {:comma? false :force-nl? true :justify? true} :vector {:wrap? false}})

[{:abcd :abcd
  :abcdef {:abcdefg  [:abcdefgh :abcdefghijkl :850fda697252 variable]
           :abcdefgh [:ab :abcdefgh-ijklmn :abcdefghijkl :850fda697252 variable]}
  :abcdefghijkl
    [{:abcd "bar"
      :abcdefg
        [:abcdefgh :abcdefghijkl :abcdefgh :850fda697252 variable]}]
  :abce {:abcd   :abcd
         :abcdef {:abcd "foo"}}}]
```
If you turn off `:force-nl?`, it will justify in a vector just fine:
```clojure
(czprint [i48z] {:width 100 :map {:comma? false :force-nl? false :justify? true} :vector {:wrap? false}})

[{:abcd         :abcd
  :abcdef       {:abcdefg  [:abcdefgh :abcdefghijkl :850fda697252 variable]
                 :abcdefgh [:ab :abcdefgh-ijklmn :abcdefghijkl :850fda697252 variable]}
  :abcdefghijkl [{:abcd "bar" :abcdefg [:abcdefgh :abcdefghijkl :abcdefgh :850fda697252 variable]}]
  :abce         {:abcd :abcd :abcdef {:abcd "foo"}}}]
```
I think a workaround until I fix this problem would be to just turn off `:force-nl?` and live with the times that maps get printed on one line.  As unpleasant as that can be.

Thank you for pointing this out.  The justification code is quite involved, and so it may be a bit before I can work this one out.  

In the process of researching this, I also found another situation where something isn't working like I would expect, so you may have exposed another bug along the way, which is good to know about! (Look at the unjustified map, above -- `[i48z]`.  The map inside of the inner vector, the `:abcdefg` key on the line after the `"bar"` value, should have its value on the same line as the `:abcdefg` key.)

I'll let you know when I get this fixed.

**Comment by mbossenbroek** [comment]
Glad you could find the issue so quickly & thanks for the workaround!

**Comment by kkinnear** [comment]
I've changed the defaults for now, and possibly permanently, to disable the particular optimization that was causing this problem.  Available in 0.4.8.
